ast.cpp

```diff
@@ -44,47 +44,97 @@ object::object(){
     offSet=++counter*8;
 }
 
+CType::~CType() {}
+
 NodeBinOp::NodeBinOp(NodeExpr* l, NodeExpr* r, string p) {
     lhs = l;
     rhs = r;
     punct = p;
     lhs->parent = this;
     rhs->parent = this;
 }
-NodeAssign::NodeAssign(NodeExpr* _lhs, NodeExpr* _rhs) : NodeBinOp(_lhs, _rhs, "=") {}
 
-NodeDiv::NodeDiv(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "/") {}
+NodeAssign::NodeAssign(NodeExpr* _lhs, NodeExpr* _rhs) : NodeBinOp(_lhs, _rhs, "=") {
+    if(lhs->is_NodeId()) {
+        lhs->c_type = rhs->c_type;
+        var_map[((NodeId*)(lhs))->id]->c_type = lhs->c_type;
+    }
+    c_type = rhs->c_type;
+}
 
-NodeSub::NodeSub(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "-") {}
+NodeDiv::NodeDiv(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "/") {
+    c_type = new CIntType();
+}
 
-NodeAdd::NodeAdd(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "+") {}
+NodeSub::NodeSub(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "-") {
+    if(rhs->c_type->isIntType() && lhs->c_type->isPtrType()) {
+        c_type = new CPtrType(new CIntType());
+    }
+    else{
+        c_type = new CIntType();
+    }
+}
 
-NodeMul::NodeMul(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "*"){}
+NodeAdd::NodeAdd(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "+") {
```

Comment:
same as NodeSub constructor except ptr + ptr should error out. it should be able to handle all 4 like this, crashing on ptr + ptr
ptr + num (this is currently implemented in this if)
ptr + ptr (intentional crash not implemented, you need an else-if for this case that crashes)
num + ptr (not implemented, you need an else-if)
num + num (currently implemented in else block)